Make Searchable Snapshot's CacheFile Lock less

original-brownbear opened the pull request:
Replacing the mechanism for eviction and listener references via a read-write lock by
a reference counting implementation.
This fixes a bug that caused test failure #63586 in which concurrently trying to acquire or release
an eviction listener while doing a file operation would sometimes lead to throwing an exception
since the `tryLock` call on the read lock would fail in this case.
Also this removes the possibility of blocking cluster state updates as a result of them waiting
on the write-lock which might take a long time if a slow read operation executes concurrently.

Closes #63586

NOTE: I ran almost 10k iterations of all searchable snapshot integ tests over this one without failure, this change also visibly speeds up tests.

original-brownbear committed:
Make Searchable Snapshot's CacheFile Lock less

Replacing the mechanism for eviction and listener references via a read-write lock by
a reference counting implementation.
This fixes a bug that caused test failure #63586 in which concurrently trying to acquire or release
an eviction listener while doing a file operation would sometimes lead to throwing an exception
since the `tryLock` call on the read lock would fail in this case.
Also this removes the possibility of blocking cluster state updates as a result of them waiting
on the write-lock which might take a long time if a slow read operation executes concurrently.

Closes #63586

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-cache-file-locking

original-brownbear committed:
nicer

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -42,37 +40,60 @@
+    /**
+     * A reference counted holder for the current channel to the physical file backing this cache file instance.
+     */
+    private static final class FileChannelReference extends AbstractRefCounted {
```
By moving access to the file channel to a ref-counted basis and giving the channel its own life-cycle, we can remove all need for locking when dealing with file closing and opening.
Background operations running for index inputs that get closed concurrently are tied to a specific instance of this reference and will simply fail once all references to the channel have been released.

original-brownbear replied:
Sure done :)

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -113,22 +117,20 @@ public boolean acquire(final EvictionListener listener) throws IOException {
-                listeners = Collections.unmodifiableSet(newListeners);
+                synchronized (listeners) {
+                    ensureOpen();
+                    final boolean added = listeners.add(listener);
```
There's no need for a `volatile` and immutable set of listeners here. We never accessed the listeners outside of the lock so we can simply mutate them in place.
I made this clearer now by using the `listeners` set as the mutex for all operations that apply to the set and the channel opening/closing that is a result of the set size.

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -140,96 +142,63 @@ public boolean release(final EvictionListener listener) {
-                }
-            } finally {
-                evictionLock.unlock();
+        if (evicted.compareAndSet(false, true)) {
```
I found this a lot cleaner and easier to reason about than checking the value of this boolean twice with tricky locking and the cost of the atomic reference relative to the `volatile` boolean should be trivial.

original-brownbear committed:
missing comment

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -498,34 +496,19 @@ public void prefetchPart(final int part) throws IOException {
                     // noinspection UnnecessaryLocalVariable
                     final Tuple<Long, Long> rangeToRead = rangeToWrite;
-
-                    try (Releasable ignored = cacheFile.fileLock()) {
```
This lock seems pointless to me and was causing trouble. If an eviction of the current file is concurrently started, all that happened here was that we would block that eviction when as far as I can tell we should simply fail the cache population as it's just wasted effort anyway as soon as the eviction hits?

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -164,193 +164,191 @@ protected void readInternal(ByteBuffer b) throws IOException {
 
         try {
             final CacheFile cacheFile = getCacheFileSafe();
-            try (Releasable ignored = cacheFile.fileLock()) {
```
Same argument as in the other spot where I removed use of this method, what's the point of locking here when all we do is make the cache file be used after eviction has already been started, meaning all the writes going into it are wasted anyway?

tlrx replied:
The point was to keep the file around in order to be able to process any concurrent read operation (if any) that would have required a gap to be filled. But there was no reason to prevent other listeners to be acquired nor an eviction to kick in. Your change improves all of that.

original-brownbear committed:
even better

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

original-brownbear commented:
@tlrx I think this is good for review now. I simplified it a little more and added some more comments. The test failure is not related to this PR but some dep download thing we currently see in master as well. Hope this makes sense :)

tlrx reviewed (approved):
LGTM - This is nice work @original-brownbear! I can't find any situations where this change could cause an issue. It simplifies the code while reducing the blocking situations to the minimum and that will also help when file are accessed within the cluster state applier thread.

Still, I'd like a second pair of eyes on this before it got merged in.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-cache-file-locking

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -306,21 +276,18 @@ private void ensureOpen() {
                     @Override
                     protected void doRun() {
                         for (SparseFileTracker.Gap gap : gaps) {
+                            ensureOpen();
```
My bad, this was a leftover from a previous attempt to fix the implementation that used locking. Will remove :)

original-brownbear committed:
CR: comments

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -42,37 +40,60 @@
+        private final FileChannel fileChannel;
+
+        FileChannelReference(Path file) throws IOException {
+            super("FileChannel");
```
done

original-brownbear commented:
Jenkins run elasticsearch-ci/1 (dep download issue ...)

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -140,96 +149,63 @@ public boolean release(final EvictionListener listener) {
+                if (listeners.isEmpty()) {
+                    // nobody is using this file so we close the channel
+                    channelRef.decRef();
+                    channelRef = null;
```
Add `assert channelRef.recCount() == 0`?

original-brownbear replied:
We can't assert that. We could still be using that specific channel in a background operation. This fact is to some degree at the heart of what motivated this PR: we don't want to lock here and wait for all the background fetchers/warmers etc. to be done with the file so we don't have to block and instead have the last of them to finish its work close it.

henningandersen replied:
Thanks, I wonder if we should then just change the comment a few lines up to not say "close" (or just remove the comment), at least that is what confused me.

henningandersen reviewed (commented):
I did an initial read-through but need a bit more time to do an in-depth review.

I wonder if it was possible to write a test that provokes the issue of #63586 with higher certainty?

original-brownbear commented:
> I wonder if it was possible to write a test that provokes the issue of #63586 with higher certainty?

I suppose technically we could construct the scenario that lead to this issue by some tricky blocking of repository reads at just the right points. But I'm not sure we want that:
1. That would be a significant effort and would only be a relevant special case in the old code, not in the new code this PR introduces.
2. We can probably make it much more likely to shake out this kind of issue and others similar to it by just more aggressively testing concurrent search + relocate which is the actual thing that caused trouble here but I'm not sure how far we want to go here. It's always a trade-off between making the tests slower and increasing the chance of catching things I guess and I'd have a hard time at guessing better parameters for parallelism etc. than what we're currently working with.

henningandersen commented:
Yeah, it is tricky. I was thinking that we could have a test of just `CacheFile`, calling `acquire`, `release`, `startEviction`, `populateAndRead` in multiple threads (in legal ways), with some level of blocking in the `reader` and `writer` passed to `populateAndRead`.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-cache-file-locking

original-brownbear committed:
better test and fix

original-brownbear commented:
@henningandersen I think I found a way of getting reasonable coverage without too much effort :) I pushed 3c3756af25d where I used the deterministic thread pool to simulate a number of possible permutations of adding and releasing a listener while concurrently evicting the file. This actually tracked down a real-world bug (that was obvious in hindsight but would only show on Windows ... or using `WindowsFS` as in this case:)) that is fixed in the same commit.
I could potentially extend that test a little to use more eviction listeners for example if you want (not sure how much that actually changes, it covers most possible logical scenarios as far as I can tell) and maybe that's all the coverage we need?

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -140,96 +156,63 @@ public boolean release(final EvictionListener listener) {
+            final Set<EvictionListener> evictionListeners;
+            synchronized (listeners) {
+                evictionListeners = new HashSet<>(listeners);
+                refCounter.decRef();
```
nit: I think this can go outside the synchronized block? Not that I have a case where it makes a difference though...

original-brownbear replied:
Right :) ... always good to not do some potentially long running file delete under a lock I guess :)

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -113,22 +131,20 @@ public boolean acquire(final EvictionListener listener) throws IOException {
         ensureOpen();
         boolean success = false;
         if (refCounter.tryIncRef()) {
```
This seems unchanged, but I find it slightly odd to have both the option of returning false from this method as well as throwing `AlreadyClosedException` when this `CacheFile` is being evicted. Am I missing something here?

original-brownbear replied:
Yea I'm not sure about this either. I think we should just always throw here, since the only way we could conceivably return `false` is if we evict between `ensureOpen()` and trying to increment the ref counter (which is close to impossible).
Maybe change that in a follow-up though, could simplify this and also the `release` method which always returns `true` currently as well and can be simplified to a `void`.

tlrx replied:
Yes, I think this is a left over when the IndexInput tried to re-acquire a cache file if it failed to add itself as a listener. I guess we were just returning `false` at that time in case of eviction.

Anyway, I'm +1 on throwing.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CacheFileTests.java:
```diff
@@ -124,9 +130,40 @@ public void testDeleteOnCloseAfterLastRelease() throws Exception {
+            random()
+        );
+        final ThreadPool threadPool = deterministicTaskQueue.getThreadPool();
+        final CompletableFuture<Integer> readFuture = cacheFile.populateAndRead(
```
Can we also (sometimes, randomly) call readIfAvailableOrPending?

original-brownbear replied:
sure thing adding it :)

henningandersen reviewed (commented):
Thanks for adding the test. This change looks great. I have a number of smaller comments only.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-cache-file-locking

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -244,7 +227,7 @@ public String toString() {
-            + (channel != null ? "yes" : "no")
+            + (channelRef != null ? "yes" : "no")
             + ", listeners="
             + listeners.size()
```
I think that's fine, it's a `volatile` field so it's not going to be totally out of whack and the behavior is the same we had so far? Not great maybe but even if we sync on `listeners` I think we won't get a completely consistent view of things as we have no lock over all the fields used here (the `evicted` can still change async) so maybe it's not worth it? (my main contention with adding locking is that it would mess with debugging in potentially by introducing a new barrier depending on log settings)

henningandersen replied:
I believe `listeners` is no longer `volatile` with this PR? So could in theory be unboundedly wrong.

Also in theory, HashSet implementation could evolve in ways that make this fail in bad ways when concurrently modifying the listeners set (though hard to imagine it would ever do so).

original-brownbear replied:
Urgh you're right I totally forgot that `listeners` isn't `volatile` any longer. I'll just `synchronize` this method on `listeners` for now as we do in all other spots where we touch `listeners`, it's just the only consistent way of doing things with the way it's implemented now IMO. It's probably not too necessary to do anything more complicated just for debugging purposes?

original-brownbear committed:
drop pointless check

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -374,6 +356,27 @@ public void onFailure(Exception e) {
+            reference = channelRef;
+            assert reference.refCount() > 0 : "impossible to run into a fully released channel reference under the listeners mutex";
+            reference.incRef();
+            future.handle((res, t) -> {
```
> I wonder if it was more intuitive to "just" ensure we decRef the FileChannelReference appropriately in the two call-sites?

Yea it's quite annoying to code this up (just adds a lot of random code) actually but let's do it for correctness :)

original-brownbear replied:
Refactored this in d5eb341a775 now to manually release the ref instead of using the future. Let me know what you think. I was able to keep it reasonably dry and clean I hope :)

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -374,6 +356,27 @@ public void onFailure(Exception e) {
+        synchronized (listeners) {
+            ensureOpen();
+            reference = channelRef;
+            assert reference.refCount() > 0 : "impossible to run into a fully released channel reference under the listeners mutex";
```
++

original-brownbear committed:
CR: comments

original-brownbear committed:
spotless

original-brownbear commented:
Thanks for the review @henningandersen I think I addressed all your comments now :) (also ran a few thousand iterations of the tests with the latest changes to releasing the file reference and it still seems to run fine.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -307,20 +289,16 @@ private void ensureOpen() {
-                                    readLock.unlock();
+                                    reference.decRef();
                                 }
+                                gap.onCompletion();
```
Looking at this again I think that `gap.onCompletion();` should be executed within the try/finally block, before the reference is decremented, as completing the gap might trigger the execution of some listeners.

original-brownbear replied:
I think we have this covered by the fact that we increment the reference by one anyway in `getFileChannelReference` so unless we are in some broken state where the overall listener got failed async (maybe that's impossible even?) the ref counting here is actually redundant.
I actually added the ref counting in `getFileChannelReference` to cover this exact spot so we're not without a reference when completing the listener. (I suppose we could do away with the incrementing in `getFileChannelReference` and decrementing after if we made this change, but it seemed easier to reason about the way it currently is but maybe that's just me?)

tlrx replied:
Yes, sorry, I've been puzzled a minute by the decRef following by the gap's completion instead of the other way around... sorry for the noise you can ignore my comment.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CacheFileTests.java:
```diff
@@ -124,9 +130,47 @@ public void testDeleteOnCloseAfterLastRelease() throws Exception {
+        final ThreadPool threadPool = deterministicTaskQueue.getThreadPool();
+        final Future<Integer> readFuture;
+        if (randomBoolean()) {
+            readFuture = cacheFile.readIfAvailableOrPending(Tuple.tuple(0L, length), channel -> Math.toIntExact(length));
```
It would be nice in this case to randomly first invoke populateAndRead too - in order to get into the more interesting part of `readIfAvailableOrPending` involving the listener callback.

original-brownbear replied:
++

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -307,20 +289,16 @@ private void ensureOpen() {
-                                if (readLock.tryLock() == false) {
-                                    throw new AlreadyClosedException("Cache file channel is being evicted, writing attempt cancelled");
+                                if (reference.tryIncRef() == false) {
+                                    throw new AlreadyClosedException("Cache file channel has been released and closed");
```
Related to your discussion below I believe we could add:
```
assert false : "expected a non-closed channel reference"
```

original-brownbear replied:
Yea that seems like a safe thing to try :) thanks!

henningandersen reviewed (approved):
LGTM, thanks @original-brownbear

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-cache-file-locking

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -349,31 +327,72 @@ public void onFailure(Exception e) {
+     *
+     * @return file channel reference
+     */
+    private FileChannelReference getFileChannelReference() {
```
++

original-brownbear committed:
CR: comments

original-brownbear commented:
Thanks Henning + Tanguy!

tlrx commented:
Thanks to you!
